jupyter_client/manager.py

```diff
@@ -584,9 +647,23 @@ async def finish_shutdown(self, waittime=None, pollinterval=0.1):
         if waittime is None:
             waittime = max(self.shutdown_wait_time, 0)
         try:
-            await asyncio.wait_for(self._async_wait(pollinterval=pollinterval), timeout=waittime)
+            try:
+                self._shutdown_status = _ShutdownStatus.ShutdownRequest
+                await asyncio.wait_for(
+                    self._async_wait(pollinterval=pollinterval), timeout=waittime / 2
+                )
+            except asyncio.TimeoutError:
+                self.log.debug("Kernel is taking too long to finish, terminating")
+                self._shutdown_status = _ShutdownStatus.SigtermRequest
+                await self._send_kernel_sigterm()
+
+            await asyncio.wait_for(
+                self._async_wait(pollinterval=pollinterval), timeout=waittime / 2
+            )
```

Comment:
I think this spin-wait needs to be associated with the `except` block - otherwise, it will occur again for the successful shutdown-request case (which will immediately exit, but probably not the intent).

Reply:
Yeah, it's with the one just below, let me try to change the layout to be clearer.

Reply:
I've split the two try/except to make the intention clearer, is that better ?